File: lib/changelog.ts
// Changelog recorder: a structured, append-only log of every change a mutating
// operation makes to the dump or the DB. Records are buffered and flushed as JSONL
// (one JSON object per line) so the file is a greppable, streamable history across
// runs. A null path makes it a no-op (the operation just doesn't log).
//
// The JSONL line schema is documented for users in README.md.

export type ChangeOp =
  | "added" // a new article file written
  | "edited" // an existing article rewritten (metadata changed)
  | "deleted" // an article removed/archived on our side (reconcile)
  | "body-added" // enrich filled a previously-empty body
  | "body-changed" // enrich replaced an existing body
  | "body-error"; // enrich recorded a bodyError instead of a body

export interface ChangeRecord {
  op: ChangeOp;
  documentType: string;
  id: string;
  title?: string;
  /** which fields changed (for op="edited"), e.g. ["metadata","updated_published"]. */
  changed?: string[];
  hashOld?: string;
  hashNew?: string;
  /** the operation that produced this record: "dump" | "enrich" | "reconcile" | "sync". */
  source?: string;
  /** free-text extra (e.g. bodyError message, archive path). */
  detail?: string;
}

// Default changelog filename, placed alongside the dump (next to _index.json).
export const CHANGELOG_BASENAME = "_changelog.jsonl";

// Resolve a `--changelog[=FILE]` flag to a path, or null when the flag is absent.
//   (absent)            -> null  (logging disabled)
//   --changelog         -> <dumpDir>/_changelog.jsonl
//   --changelog=FILE    -> FILE
export function changelogPathFromFlag(
  flagValue: string | boolean | undefined,
  dumpDir: string,
): string | null {
  if (flagValue === undefined) return null;
  if (flagValue === true || flagValue === "") {
    return `${dumpDir.replace(/\/+$/, "")}/${CHANGELOG_BASENAME}`;
  }
  return String(flagValue);
}

export class Changelog {
  private buf: string[] = [];
  private counts: Record<string, number> = {};

  /** path=null disables logging; runId labels every record from this run. */
  constructor(private readonly path: string | null, private readonly runId: string) {}

  get enabled(): boolean {
    return this.path !== null;
  }

  record(rec: ChangeRecord): void {
    this.counts[rec.op] = (this.counts[rec.op] ?? 0) + 1;
    if (!this.path) return; // count even when not persisting, for summaries
    this.buf.push(
      JSON.stringify({ runId: this.runId, ts: new Date().toISOString(), ...rec }),
    );
  }

  /** Per-op tally (works even when disabled, so callers can still summarize). */
  byOp(): Record<string, number> {
    return { ...this.counts };
  }

  get total(): number {
    return Object.values(this.counts).reduce((a, b) => a + b, 0);
  }

  /** Append buffered records to the JSONL file (creating it if needed). No-op if
   *  disabled or nothing buffered. */
  async flush(): Promise<void> {
    if (!this.path || this.buf.length === 0) return;
    const text = this.buf.join("\n") + "\n";
    await Deno.writeTextFile(this.path, text, { append: true });
    this.buf = [];
  }
}

File: f5kb/lib/changelog.py
"""Changelog recorder: structured, append-only log of every change a mutating
operation makes to the dump or DB. Records buffered and flushed as JSONL
(one JSON object per line). A null path makes it a no-op.
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Literal

ChangeOp = Literal[
    "added",
    "edited",
    "deleted",
    "body-added",
    "body-changed",
    "body-error",
]

CHANGELOG_BASENAME = "_changelog.jsonl"


def changelog_path_from_flag(flag_value: str | bool | None, dump_dir: str) -> str | None:
    """Resolve a --changelog[=FILE] flag to a path, or None when absent.

      (absent)         -> None  (logging disabled)
      --changelog      -> <dumpDir>/_changelog.jsonl
      --changelog=FILE -> FILE
    """
    if flag_value is None:
        return None
    if flag_value is True or flag_value == "":
        base = str(dump_dir).rstrip("/")
        return f"{base}/{CHANGELOG_BASENAME}"
    return str(flag_value)


class Changelog:
    def __init__(self, path: str | None, run_id: str) -> None:
        self._path = path
        self._run_id = run_id
        self._buf: list[str] = []
        self._counts: dict[str, int] = {}

    @property
    def enabled(self) -> bool:
        return self._path is not None

    def record(self, op: ChangeOp, document_type: str, id: str, **kwargs: object) -> None:
        self._counts[op] = self._counts.get(op, 0) + 1
        if not self._path:
            return
        entry: dict = {
            "runId": self._run_id,
            "ts": _iso_now(),
            "op": op,
            "documentType": document_type,
            "id": id,
            **kwargs,
        }
        self._buf.append(json.dumps(entry))

    def by_op(self) -> dict[str, int]:
        return dict(self._counts)

    @property
    def total(self) -> int:
        return sum(self._counts.values())

    def flush(self) -> None:
        if not self._path or not self._buf:
            return
        text = "\n".join(self._buf) + "\n"
        p = Path(self._path)
        p.parent.mkdir(parents=True, exist_ok=True)
        with p.open("a", encoding="utf-8") as f:
            f.write(text)
        self._buf.clear()


def _iso_now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + "Z"
